packages/core/src/typesnew.ts

```diff
@@ -0,0 +1,213 @@
+import { ElementDefinition, ResourceType, StructureDefinition } from '@medplum/fhirtypes';
+import { TypedValue } from './types';
+import { getTypedPropertyValue } from './fhirpath';
+
+/**
+ * Internal representation of a non-primitive FHIR type, suitable for use in resource validation
+ */
+export interface InternalTypeSchema {
+  name: ResourceType;
+  fields: Record<string, FieldValidator>;
+  constraints: Constraint[];
+}
+
+export interface FieldValidator {
```

Comment:
If we're introducing new types, it might be a nice opportunity to be more disciplined on aligning with FHIR naming, so for example calling this `ElementValidator`?